doc/man/man1/correl.1
\"    This documentation is copyright 1994 David Huron.
.TH correl 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +12
.ti -12
\fBcorrel\fR  --  measure the numerical similarity between two Humdrum spines
.in -12
.in -2
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
\fBcorrel\fR  [\fB-f \fItemplatefile\fR]  [\fB-m\fR]  [\fB-p \fIn\fR]  [\fB-s \fIregexp\fR]  [\fIinputfile ...\fR]  [> \fIoutputfile.cor\fR]
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "correl"
command measures the degree of parametric (numerical) similarity
between corresponding values in two Humdrum spines.
More precisely,
.B "correl"
calculates Pearson's coefficient of correlation for
paired tokens containing numerical data.
.sp 1
.sp 1
Two modes of operation are provided.
In the
.I "single input mode,"
a single file containing two equal-length spines is processed.
In this mode, the output from
.B "correl"
consists of a single number indicating the linear correlation between
the two spines of numerical data.
.sp 1
.sp 1
In the
.I "dual input mode,"
two single-spine numerical inputs -- called the
.I "template file"
and the
.I "input file"
-- are specified by the user.
Normally, the template file is considerably shorter than the input file.
In this mode, the output consists of a spine of numerical information
(\f(CR**correl\fR) that reflects the momentary similarity between
the template and the input for each successive moment in the input.
In short, the input file is `scanned' using the template values,
and the correlational similarity determined at each point.
.sp 1
.sp 1
In both
.I "single input"
and
.I "dual input"
modes, output numerical values range between +1 and -1.
Correlation values reflect the degree to which two sets of
numerical values rise and fall in synchrony.
The maximum output value is +1 -- indicating that the two sets
of numbers are perfectly related according to a linear relationship.
A minimum output value of \(mi1 indicates that the two sets of numbers
are perfectly out-of-phase -- one set of numbers rises while
the other set falls by a proportional magnitude, and vice versa.
A correlation value of zero indicates that there is no linear
relationship between the two sets of numbers.
.sp 1
.sp 1
In
.I "single input mode,"
inputs to
.B "correl"
must consist of precisely two spines;
otherwise an error message is generated and the command is terminated.
The two spines may contain different interpretations and represent
different types of information.
In the case of the
.I "dual input mode,"
the input file and template file must have precisely one spine each;
the spines may differ in length, but the
.I "template"
file must not be longer than the
.I "input"
file.
.sp 1
.sp 1
Only numerical signifiers are considered by
.B "correl;"
non-numeric input data are ignored.
Where a data token contains a mix of numeric and non-numeric signifiers,
only the first complete numerical subtoken contributes to the calculation.
The following examples illustrate how
.B "correl"
interprets mixed data tokens:
.in +2
.sp 1
.TS
c c
l l.
data token	numerical interpretation

\f(CR4gg#	\fR4
\f(CR4.gg#	\fR4
\f(CR-33aa	\fR-33
\f(CR-aa33	\fR33
\f(CRx7.2yz	\fR7.2
\f(CRa7..2bc	\fR7
\f(CR[+5]12	\fR5
\f(CR$17@2	\fR17
\f(CRa1b2 c.3.d	\fR1   0.3
.TE
.sp 1
.in -2
Humdrum multiple-stops require special attention in
.B "correl"
(see below).
.sp 1
.sp 1
In the
.I "dual input mode"
the output from the
.B "correl"
command consists of a set of records matching the structure of the
.I "input"
document.
Output values indicate the correlation between the
.I "template"
data values and the input data values beginning at that record.
.sp 1
.sp 1
When the
.I "dual input mode"
is invoked, it is recommended that output files produced by the
.B "correl"
command should be given names with the distinguishing `.cor' extension.
.in -2
.sp 1
.sp 1
.SH "OPTIONS"
.in +2
The
.B "correl"
command provides the following options:
.sp 1
.TS
l l.
\fB-f \fItemplatefile\fR	specify source pattern as \fItemplatefile\fR
	and invoke dual input mode
\fB-h\fR	displays a help screen summarizing the command
	syntax
\fB-m\fR	disable matched-pairs criterion
\fB-p \fIn\fR	output precision to \fIn\fR decimal places
\fB-s \fIregexp\fR	skip; completely ignore data records matching
	\fIregexp\fR
.TE
.sp 1
Options are specified in the command line.
.sp 1
.sp 1
The
.B "-f"
option is used to specify an independent
.I "template file"
and so invoke the
.I "dual input mode."
.sp 1
.sp 1
The
.B "-p"
option can be used to set the precision of the output values to
.I "n"
decimal places.
The default precision is 3 decimal places.
.sp 1
.sp 1
The
.B "-s"
option allows the user to avoid (or skip) the processing of certain
types of data records.
This option must be accompanied by a user-defined regular-expression.
Input data records matching this expression are not processed.
.sp 1
.sp 1
Correlation values can be calculated only where all numerical data
are arranged as matched pairs -- that is, the input conforms to
the \(odmatched pairs criterion.\(cd
For example, the following two spines illustrate numerical data matching.
The number of numerical data values in both spines are matched throughout
the inputs:
.in +2
.sp 1
.TS
l l.
**spine1	**spine2
10.0	4
7 3	2 .91
\.	.
13.8	4
5 8 5	1 1 2
a b c	x
\.	p q
*-	*-
.TE
.sp 1
.in -2
By contrast, the following file shows several transgressions of
the matched pairs criterion.
For example, the first data record gives a numerical value in spine #1
that is not matched by a numerical value in spine #2.
Similarly, the multiple-stop values in the second data record
are unmatched in spine #2:
.in +2
.sp 1
.TS
l l.
**spine1	**spine2
9.7	a
7 31	2
\.	114
426	.
r 11 7	35 xy08z 28
a b c	6 .07
\.	p q
*-	*-
.TE
.sp 1
.in -2
In normal operation, a single failure to conform to
the matched pairs criterion will cause
.B "correl"
to issue an error message and terminate operation.
If the
.B "-m"
option is invoked, unmatched data is simply ignored.
For example, with the
.B "-m"
option, the above input is treated as equivalent to the
following input:
.in +2
.sp 1
.TS
l l.
**spine1	**spine2
\.	.
7	2
\.	.
\.	.
11 7	35 08
\.	.
\.	.
*-	*-
.TE
.sp 1
.in -2
.in -2
.SH "EXAMPLES"
.in +2
The following examples illustrate the operation of
.B "correl."
The first example shows an excerpt containing considerable
parallel motion between two polyphonic voices.
Measuring the pitch-contour similarity can be done using the
single input mode.
.in +2
.sp 1
.TS
l s s s
l l l l.
!! J.S. Bach, Invention No. 8; BWV 779
**semits	**semits
*M3/4	*M3/4
9	17
12	21
10	19
12	21
9	17
12	21
10	19
12	21
\(eq6	\(eq6
5	14
9	17
7	16
9	17
*-	*-
.TE
.sp 1
.in -2
In order to avoid processing the measure numbers, the skip (\fB-s\fR)
option is used; executing the command:
.sp 1
.sp 1
.in +2
correl -s = bwv779
.in -2
.sp 1
.sp 1
will produce the following output:
.sp 1
.sp 1
.in +2
0.979
.in -2
.sp 1
.sp 1
The second example illustrates the dual input mode.
The target input consists of a single spine (labelled
**input)
containing mixed alphabetic and numerical values.
(This input file is shown below as the left-most spine.)
The template file consists of the numerical sequence:
1, 2, 3 -- mixed with the letters a, b, c.
(This file is shown as the middle spine below.)
Note that the non-numeric characters in both the input
and template files have no influence on the operation of
.B "correl."
The third (output) spine is produced by the following command:
.sp 1
.sp 1
.in +2
correl -f template input > output.cor
.in -2
.sp 1
.sp 1
.in +2
.sp 1
.TS
c c c
c c c
l l l.
(input	(template	(correl
file)	file)	output)
**input	**template	**correl
0	1a	1.000
1	2b	1.000
2	3c	1.000
3	*-	-0.655
4		-0.655
x1x		0.866
y2.		0.866
2z		0.000
(3)		-1.000
[2]		.
01		.
*-		*-
.TE
.sp 1
.in -2
The similarity values generated by
.B "correl"
are given in the
**correl
spine.
Each successive value in the output spine is matched with a data token in
the target input file (\f(CR**foo\fR).
For example, the initial three output values (1.000)
indicate that exact positive correlations occur between the template
and the input.
That is (0, 1, 2) (1, 2, 3) and (2, 3, 4) all show simple equi-distant
increases corresponding to the source template.
The final numerical value in
**correl
shows a negative correlation (-1.000) indicating that the numerical sequence
(3, 2, 1) is the exact opposite contour to the source template (1, 2, 3).
By contrast, the immediately preceding output
value (0.000) indicates that the sequence (2, 3, 2)
shows no systematic linear relationship with the source template (1, 2, 3).
.sp 1
.sp 1
The following example provides a more complicated illustration of
.B "correl."
Once again the left-most spine is the target input,
the middle spine is the source template, and the right-most spine
shows the corresponding output.
.in +2
.sp 1
.TS
c c c
c c c
l l l.
(input	(template	(correl
file)	file)	output)
**input	**template	**correl
\(eq1	1	.
1	2 3	1.000
2 3	.	-0.370
100	4	-0.742
8r	5 6	.
4	*-	0.042
5 6		.
\(eq2		.
0		.
4r		.
-2x -3		.
-x8		.
\(eq\(eq		.
*-		*-
.TE
.sp 1
.in -2
The above output spine was created by executing the command:
.sp 1
.sp 1
.in +2
correl -m -s '[=r]' -f template input > output.cor
.in -2
.sp 1
.sp 1
Due to the
.B "-s"
option, all records in the input file containing an equals-sign
or lower-case `r' are eliminated from the calculations.
The presence of the null-token in the third data record of the template file is
noteworthy.
Although no correlations are calculated with the null-token,
it acts as a place-holder, and causes the corresponding record
in the input file to be ignored.
For example, the first correlation value is calculated on the basis
of the following coordination of numerical data:
.in +2
.sp 1
.TS
l l.
1	1
2 3 	2 3
100	.
4	4
5 6	5 6
.TE
.sp 1
.in -2
Since the value `100' is not matched with a numerical value in the
template, it is ignored in the correlation measure.
(Note that without the
.B "-m"
option, no output would be generated.)
.sp 1
.sp 1
At the next instant, the correlation value is calculated on the basis
of the following coordination of numerical data:
.in +2
.sp 1
.TS
l l.
2 3	1
100 	2 3
4	.
5 6	4
0	5 6
.TE
.sp 1
.in -2
The double-stops do not form matched pairs, hence much of the data
is discarded.
For example, in the first data record, 2 is matched with 1 and 3 is discarded.
In the second record, 100 is matched with 2 and 3 is discarded, etc.
.sp 1
.sp 1
The third correlation value is calculated on the basis
of the following coordination of numerical data:
.in +2
.sp 1
.TS
l l.
100	1
4 	2 3
5 6	.
0	4
-2 -3	5 6
.TE
.sp 1
.in -2
In this case, the correlation value is based on the following numerical pairing:
100 \(io 1, 4 \(io 2, 0 \(io 4, \(mi2 \(io 5, \(mi3 \(io 6.
All other numerical values are ignored.
.sp 1
.sp 1
The final correlation value in this example is calculated on the basis
of the following coordination of numerical data:
.in +2
.sp 1
.TS
l l.
4	1
5 6 	2 3
0	.
-2 -3	4
8	5 6
.TE
.sp 1
.in -2
The corresponding correlation value is based on the following numerical pairing:
4 \(io 1, 5 \(io 2, 6 \(io 3, \(mi2 \(io 4, 8 \(io 5.
.in -2
.sp 1
.sp 1
.SH "PORTABILITY"
.in +2
\s-1DOS\s+1 2.0 and up, with the \s-1MKS\s+1 Toolkit.
\s-1OS/2\s+1 with the \s-1MKS\s+1 Toolkit.
\s-1UNIX\s+1 systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fBpatt\fR (1), \fBpattern\fR (1), \fBsimil\fR (1)
.in -2
.sp 1
.sp 1
.SH "WARNINGS"
.in +2
Correlation coefficients indicate only the
magnitude of the association between two sets of data.
High correlation values can occur purely by chance.
The noteworthiness (statistical significance) of a
correlation value depends on the number of input values
given in the template spine.
Novice users should consult a standard statistics textbook
for further advice on how to interpret the results.
.sp 1
.sp 1
For formal statistical measures, the
.B "-m"
option should never be invoked.
.sp 1
.sp 1
If only one pair of matched values is present,
the linear correlation is mathematically undefined.
In this case a question mark signifier is output.
.in -2
.sp 1
.sp 1
.SH "LIMITS"
.in +2
The
.B "correl"
command is currently unable to handle input files greater than
about 4,000 records.
.in -2


toolkit/awk/correl.awk
#####################################################################
##                          CORREL.AWK
# Originally programmed by David Huron, 1992.
# Re-programmed by:  Tim Racinsky     Date: June, 1994
#
# Modifications:
#   Date:      Programmer:    Description:
# 1995 May 29  David Huron    Fixed divide by zero when input values are sparse.
# 1996 Sept.   Paul von Hippel Added ability to handle exponent input notation.
#
# The purpose of this program is to calculate the numerical 
# correlation between two spines.
#
# Main Functions used:
#	parse_command()			read_template()		process_comments()
#	process_interpretations()	process_dual()			process_single()
#	compute_correl()			determine_numbers()
#
#	VARIABLES:
#
#  -stderr:  variable holding destination of error messages
#  -precision:  holds precision of output specified by user (default 3)
#  -options:  holds options specified by the user
#  -template_max:  holds maximum number of data records in template array
#  -dual:  flag to indicate if dual or single input mode
#  -no_interpretation_yet:  flag to indicate if exclusive interp has been found
#  -buffer_head:  pointer to latest entry to buffer
#  -buffer_tail:  pointer to oldest, non-printed entry in buffer
#  -data_count:  holds number of data records still to be processed 
#			  currently stored in buffer
#  -error:  flag to indicate if error has occurred
#  -sin_count:  counts number of data records for single input mode
#  -precision_reg:  regular expression for precision value
#  -number_reg:  regular expression for valid number to be processed
#  -spine_terminate:  regular expression for spine terminating record
#
#	ARRAYS:
#
#  -template:  Holds data records from template file (-f option)
#  -buffer:  Holds data records from input file (-f option)
#  -sin_value1:  Holds spine one data records for single input mode
#  -sin_value2:   Holds spine two data records for single input mode
#
BEGIN {
	USAGE="\nUSAGE: correl -h                   (Help Screen)\n       correl "\
		 "[-f templatefile ] [-m] [-p n] [-s regexp] [file]"
	FS = OFS = "\t"
	TRUE = 1; FALSE = 0
	#
	# Set standard error to the appropriate 'file'
	#
	#if (ARGV[0] == "nawk") stderr = "/dev/tty"
	#else stderr = "con"
	"echo $CON" | getline stderr
	close("echo $CON")
	#
	# Set regular expressions to be used in program
	#
	spine_terminate = "^\\*-(\t\\*-)*$"
	precision_reg = "^[1-9][0-9]*$"
	number_reg = "[+-]?(([0-9]*\\.[0-9]+)|([0-9]+))"
	#
	# Parse command line options
	#
	no_interpretation_yet = TRUE
	buffer_head = data_count = 0
	buffer_tail = 1
	sin1_count = sin2_count = 0
	options = ""
	precision = 3
	skip_reg = ""
	template_max = 0
	dual = FALSE
	error = FALSE
	parse_command()
}
{
#
# Dual input mode requires a single spine in the input file
#
if (dual && NF != 1)
	{
	print "correl: ERROR: Input file must contain exactly one spine." > stderr
	error = TRUE
	exit
	}
#
# Single input mode requires two spines in the input file
#
if (!dual && NF != 2 && $0 !~ /^!!/)
	{
	print "correl: ERROR: Input file must contain exactly two spines." > stderr
	error = TRUE
	exit
	}
#
# Process records according to their types
#
if ($0 ~ /!/) process_comments()
else if ($0 ~ /^\*/) process_interpretations()
else if (options ~ /s/ && $0 ~ skip_reg)
	{
	if (dual) buffer[++buffer_head] = $0
	}
else if (dual) process_dual()
else process_single()
}
END	{
	#
	# If there are records still to be output in dual mode, print them
	#
	if (!error && dual)
		{
		for (i = buffer_tail; i < buffer_head; i += 1)
			{
			if (buffer[i] ~ /^(\*|!)/) print buffer[i]
			else print "."
			}
		#
		# Print a spine-path terminator at the end of the input regardless
		# of what is there already
		#
		if (buffer_head != 0)
			{
			if (buffer[buffer_head] == "*") print "*-"
			else if (buffer[buffer_head] ~ /^!/) print buffer[buffer_head]
			else
				{
				print "."
				print "*-"
				}
			}
		}
	#
	# If single mode, calculate correlation from stored data
	#
	else if (!error && !dual)
		{
		if (sin1_count == 0) exit
		for (i = 1; i <= sin1_count; i += 1)
			{
			sum1 += sin_value1[i]
			sum2 += sin_value2[i]
			}
		mean1 = sum1 / sin1_count
		mean2 = sum2 / sin1_count
		numerator = denominator = temp1 = temp2 = 0
		for (i = 1; i <= sin1_count; i += 1)
			{
			numerator += (sin_value1[i] - mean1) * (sin_value2[i] - mean2)
			temp1 += (sin_value1[i] - mean1)^2
			temp2 += (sin_value2[i] - mean2)^2
			}
		denominator = sqrt(temp1 * temp2)
		if (denominator == 0) print "0"
		else printf("%." precision "f\n",numerator / denominator)
		}
	}

############################################################################
#				FUNCTIONS USED IN CORREL.AWK
############################################################################

############################################################################
##				Function Parse_command
# This function inspects the array ARGV in order to determine the correct
# options that the user has specified.
#
function parse_command()
	{
	#
	# Otherwise, process any necessary variables
	#
	if (ARGV[2] != "null") options = ARGV[2]
	#
	# Obtain the precision value if specified
	#
	if (ARGV[3] != "null")
		{
		if (ARGV[3] !~ precision_reg)
			{
			print "correl: ERROR: Invalid precision specified on command "\
				 "line: " ARGV[3] > stderr
			print USAGE > stderr
			error = TRUE
			exit
			}
		else precision = ARGV[3]
		}
	#
	# Store the skip option and value if specified
	#
	if (ARGV[4] != "null")
		{
		skip_reg = ARGV[4]
		options = options "s"
		}
	#
	# Read in template file if specified
	#
	if (ARGV[5] != "null")
		{
		read_template(ARGV[5])
		dual = TRUE
		}
	ARGV[1] = ARGV[2] = ARGV[3] = ARGV[4] = ARGV[5] = ""
	}

############################################################################
##					Function read_template
#	This function reads in the template file if specified on the command
# line with -f option and stores the data records in the array template
#
function read_template(template_file     ,number_found)
	{
	number_found = FALSE
	#
	# Loop through the template file
	#
	while (getline < template_file > 0)
		{
		#
		# The template file must have only one spine of input
		#
		if (NF != 1)
			{
			print "correl: ERROR: Must specify precisely one spine in the "\
				 "template file." > stderr
		 	error = TRUE
			exit
			}
		#
		# Skip any interpretations, commens or user defined skips
		#
		else if (($0 ~ /^(\*|!)/) || (options ~ /s/ && $0 ~ skip_reg))
			continue
		else
			{
			template[++template_max] = $0
			if (!number_found && $0 ~ number_reg) number_found = TRUE
			}
		}
	#
	# The template array must have at least one data record containing a number
	#
	if (!number_found)
		{
		print "correl: ERROR: Must specify at least one valid data record "\
			 "in template file." > stderr
		exit
		}
	close(template_file)
	}

###########################################################################
##				Function process_comments
#	This function stores comment records if dual input mode is specified
#
function process_comments()
	{
	if (dual) buffer[++buffer_head] = $0
	}

############################################################################
##				Function process_interpretations
#	This function processes interpretation records if dual input mode
#
function process_interpretations()
	{
	if (dual)
		{
		#
		# Store the exclusive interpretation to show the type of output
		#
		if (no_interpretation_yet)
			{
			no_interpretation_yet = FALSE
			buffer[++buffer_head] = "**correl"
			}
		else
			{
			#
			# The only types of interpretations that should be stored
			# are spine terminators
			#
			#if ($0 ~ spine_terminate)
			#	{
			#	buffer[++buffer_head] = "*-"
			#	no_interpretation_yet = TRUE
			#	}
			#else buffer[++buffer_head] = "*"
			buffer[++buffer_head] = "*"
			}
		}
	}

#########################################################################
##				Function process_dual
#	This function processes data records if dual input mode is specified
#
function process_dual(   i)
	{
	#
	# Add the data record to the buffer array
	#
	buffer[++buffer_head] = $0
	data_count += 1
	#
	# If there are enough data records accumulated in the buffer array, then
	# compute the correlation
	#
	if (data_count == template_max)
		{
		i = buffer_tail
		#
		# Loop through and print off any records that will not be computed
		# in the correlation
		#
		while (buffer[i] ~ /^(\*|!)/ || \
								(options ~ /s/ && buffer[i] ~ skip_reg))
			{
			if (buffer[i] ~ /^(\*|!)/) print buffer[i]
			else print "."
			buffer_tail += 1
			i += 1
			}
		#
		# Compute the correlation at the current record and modify the
		# variables accordingly
		#
		compute_correl()
		buffer_tail += 1
		data_count -= 1
		}
	}

###########################################################################
##				Function compute_correl
#	This function performs the actual correlation computation
#
function compute_correl(    i,temp_count,no_of_values,sum1,sum2,mean1,mean2,\
				  temp_buf,numerator,denominator,temp1,temp2,value1,value2)
	{
	#
	# Place into a temporary array the records in buffer that should be
	# processed (i.e. only data records)
	#
	temp_count = 0
	for (i = buffer_tail; i <= buffer_head; i += 1)
		{
		if ((buffer[i] ~ /^(\*|!)/)||(options ~ /s/ && buffer[i] ~ skip_reg))
			continue
		else
			temp_buf[++temp_count] = buffer[i]
		}
	#
	# Place the numbers to be processed into the arrays value1 and value2
	#
	no_of_values = determine_numbers(temp_buf,value1,value2)
	#
	# If the number of values is zero or one, then the correlation
	# is undefined.
	if (no_of_values <= 1)
		{
		if (denominator == 0) print "?"
		return
		}
	#
	# Compute the correlation of the values (similar to END function)
	#
	for (i = 1; i <= no_of_values; i += 1)
		{
		sum1 += value1[i]
		sum2 += value2[i]
		}
	mean1 = sum1 / no_of_values
	mean2 = sum2 / no_of_values
	numerator = denominator = temp1 = temp2 = 0
	for (i = 1; i <= no_of_values; i += 1)
		{
		numerator += (value1[i] - mean1) * (value2[i] - mean2)
		temp1 += (value1[i] - mean1)^2
		temp2 += (value2[i] - mean2)^2
		}
	denominator = sqrt(temp1 * temp2)
	if (denominator == 0) print "?"
	else printf("%." precision "f\n",numerator / denominator)
	}

############################################################################
##				Function determine_numbers
#	This function determines the numbers to be compared by looping through
# the arrays as well as any multiple stops
#
function determine_numbers(tempbuf,value1,value2,     i,j,value1_count,\
					value2_count,count1,count2,double1,double2,buf1,buf2,\
			max_count,max_number,number_found1,number_found2,number1,number2)
	{
	#
	# Loop through the given arrays (each should have the same number of
	# elements since they are the same number of data records)
	#
	for (i = 1; i <= template_max; i += 1)
		{
		count1 = count2 = 0
		#
		# Determine the number of 'number signifiers' present in the double 
		# stop of each of the array elements
		#
		double1 = split(template[i],buf1," ")
		for (j = 1; j <= double1; j += 1)
			{
			if (buf1[j] ~ number_reg) count1 += 1
			}
		double2 = split(tempbuf[i],buf2," ")
		for (j = 1; j <= double2; j += 1)
			{
			if (buf2[j] ~ number_reg) count2 += 1
			}
		#
		# If there are not the same number of 'number tokens' in the
		# current elements and -m option was no specified, print an error
		#
		if (count1 != count2 && options !~ /m/)
			{
			print "correl: ERROR: Matched-pairs criterion transgressed"\
				 " near line " FNR "." > stderr
			error = TRUE
			exit
			}
		#
		# Determine max number of multiple stops
		#
		if (double1 > double2) max_count = double1
		else max_count = double2
		#
		# Determine how many 'number tokens' should be processed
		#
		if (count1 > count2) max_number = count2
		else max_number = count1
		number_found1 = number_found2 = 0
		#
		# Loop through the multiple stops and store any 'number' tokens
		# in the arrays value1 and value2
		#
		for (j = 1; j <= max_count; j += 1)
			{
			if ((match(buf1[j],number_reg) > 0) \
									 && (number_found1 < max_number))
				{
				number1 = substr(buf1[j],RSTART,RLENGTH) + 0
				value1[++value1_count] = number1
				number_found1 += 1
				}
			if ((match(buf2[j],number_reg) > 0) \
									 && (number_found2 < max_number))
				{
				number2 = substr(buf2[j],RSTART,RLENGTH) + 0
				value2[++value2_count] = number2
				number_found2 += 1
				}
			}
		}
	# 
	# Return the number of values to be used in the computation
	#
	return value1_count
	}

##########################################################################
##				Function process_single
#	This function processes data records for single input mode.  It
# performs the same task as the function determine_numbers().
#
function process_single(     count1,count2,double1,doubl2,buf1,buf2,max_count,\
					max_number,number_found1,number_found2,number1,number2)
	{
	count1 = count2 = 0
	#
	# Determine the number of 'number signifiers' present in the double 
	# stop of each of the array elements
	#
	double1 = split($1,buf1," ")
	for (j = 1; j <= double1; j += 1)
		{
		if (buf1[j] ~ number_reg) count1 += 1
		}
	double2 = split($2,buf2," ")
	for (j = 1; j <= double2; j += 1)
		{
		if (buf2[j] ~ number_reg) count2 += 1
		}
	#
	# If there are not the same number of 'number tokens' in the
	# current elements and -m option was no specified, print an error
	#
	if (count1 != count2 && options !~ /m/)
		{
		print "correl: ERROR: Matched-pairs criterion transgressed"\
			 " near line " FNR "." > stderr
		error = TRUE
		exit
		}
	#
	# Determine max number of multiple stops
	#
	if (double1 > double2) max_count = double1
	else max_count = double2
	#
	# Determine how many 'number tokens' should be processed
	#
	if (count1 > count2) max_number = count2
	else max_number = count1
	sin_count += max_number
	number_found1 = number_found2 = 0
	#
	# Loop through the multiple stops and store any 'number' tokens
	# in the arrays sin_value1 and sin_value2
	#
	for (j = 1; j <= max_count; j += 1)
		{
		if ((match(buf1[j],number_reg) > 0) \
								 && (number_found1 < max_number))
			{
			number1 = substr(buf1[j],RSTART,RLENGTH) + 0
			sin_value1[++sin1_count] = number1
			number_found1 += 1
			}
		if ((match(buf2[j],number_reg) > 0) \
								 && (number_found2 < max_number))
			{
			number2 = substr(buf2[j],RSTART,RLENGTH) + 0
			sin_value2[++sin2_count] = number2
			number_found2 += 1
			}
		}
	}
